jetty-util/src/main/java/org/eclipse/jetty/util/thread/ThreadPoolBudget.java

```diff
@@ -127,20 +129,40 @@ public Lease leaseTo(Object leasee, int threads)
     {
         Leased lease = new Leased(leasee,threads);
         allocations.add(lease);
-        check();
+        try
+        {
+            check();
+        }
+        catch(IllegalStateException e)
+        {
+            lease.close();
```

Comment:
This `close()` here is weird.

Line:

`allocations.add(lease);`

should be done after `check()`.

Also, looking at the implementation of `close()` I don't like that `info` is being modified from within an iteration over a stream, with an unnecessary check for `contains()` that is not needed because it's a `Set`.

I'll work on some cleanup.